### ISSUE TITLE
Passing golang variables to javascript

### ISSUE THREAD

**Issue by farhank3389** [a-h/templ#944]
I have a templ block defined like this:

```
templ test(teststring string) {
    <script>
        console.log({ "the value of teststring is " + teststring });
    </script>
}

```

However this does not work. The HTML it generates looks like this:
```

    <script>
        console.log({ "the value of teststring is " + teststring });
    </script>
```

I'm not sure if I'm doing something completely wrong here as this is my first time using templ. It seems to me that something like this should work. I had a look at the `Using JavaScript with templ` docs page and it says to not use the `script` template method of doing this, which does work for me. I'm not sure how I can do the same thing using script tags as the docs say.

**Comment by pharrisee** [comment]
There is some documentation on passing server side data to client scripts:

https://templ.guide/syntax-and-usage/script-templates#pass-server-side-data-to-the-client-in-a-script-element

I also use a slightly different pattern on occasion, but tbh I'm not too happy with either way of doing it and much preferred the older script templates (https://templ.guide/syntax-and-usage/script-templates#script-templates):

```golang
templ test(teststring string) {
    <script teststring={teststring}>
        const teststring=document.currentScript.getAttribute("teststring");
        console.log({ "the value of teststring is " + teststring });
    </script>
}
```

**Comment by farhank3389** [comment]
ah yeah that does look like what I need. And agreed, I prefer the `script` template method too.

**Comment by a-h** [comment]
I can see why you like the script templates. There might be some middle ground to hit in the future, but need to control the scope of what we're working on in the short term - i.e. need to prioritise, and I think scripts are ... OK.

The problem with the script templates is really around how the initial design has boxed us in a bit. I don't think there's a clear way for us to expand on them to do async JS templates, add typescript, introduce formatting or LSP features etc. so it's more of a pause on progressing that feature direction than anything else.

Although... in August we did add a pretty cool new feature to it: https://github.com/a-h/templ/commit/ef4dde62b8cd6e53ed55b15b7443f292cc8d4d76

The need to pass variables from the Go world to the JS world won't go away, but I was looking at introducing templ variables directly into `<script>` tags instead. Combined with `templ.Once` which limits rendering of the block to once per HTTP request, it would provide all the same features, but fit better into the ecosystem.

However, I would need to find a way to slot Go variables into the JS code, by parsing the content, or using a helper function.

Just typing this out has put a new idea in my head.

```templ
type Data struct {
  X int `json:"x"`
  Y int `json:"y"`
}

templ Component(d Data) {
  <script type="text/javascript" data-templ-jsvar-name={ d }>
    alert(name.x * name.y)
  </script>
}
```

If there's a `data-templ-jsvar-*` attribute, templ could generate a JS closure that includes the `-name` as a variable, e.g.:

```html
  <script type="text/javascript">
    let name = { x: 1, y: 3 };
    alert(name.x * name.y)
  </script>
```

However, that would mean that the variable wouldn't exist in the dev environment while you're coding, which might trigger linters, since they wouldn't know how to resolve the values.

templ could generate a JS file containing type definitions (classes?) for the data, e.g.: if you add an attribute for `data-templ-jsvar-name` it takes the name of the Go type (`Data`) and creates a JS class that matches in another file.

```js
class Data {
  constructor(o) {
    //TODO: Get the data from a JSON variable in the attribute of the parent script tag.
    this.x = x;
    this.y = y;
  }
}
```

```templ
templ Component(d Data) {
  <script type="text/javascript" data-templ-jsvar-name={ d }>
    let d = new Data(this); // The Data class was auto-generated by templ.
    alert(d.x * d.y)
  </script>
}
```

Thoughts? @joerdav?

**Comment by gotzmann** [comment]
> Just typing this out has put a new idea in my head.

Looks cool! Good to have the feature, just shorten the name of tag maybe? Like `data-templ-let` or something like this?

**Comment by a-h** [comment]
Stability is important. Would never want to break people's code.

**Comment by a-h** [comment]
Also, if we _really_ wanted to remove something, we'd likely add an automatic migration feature, plus keep the old method around for at least two Go release cycles.

**Comment by joerdav** [comment]
And also, I think we would want to make sure that it was desireable to migrate a way. The alternative should be a smoother experience, I don't think we want to force a move to a unfinished solution. Once we have a proper solution hopefully people will want to move to it because it is easier.

**Comment by dimmerz92** [comment]
I might just be oversimplifying in my head or not understanding the issues you are facing with the implementation, but could we perhaps just have `script` and `async` in templ?

I personally really like the templ `script` implementation as it is. Rather than shoving a bunch of functionality into `script`, just leave it as is for simple functions or sequences of instructions and add new constructs for other features like `async` or even `typescript`?

Maybe also just a hot take on my end or just ignorance, I feel like typescript isn't really a feature that is particularly useful.. I imagine most of us are just using scripts to ship js to the front end?

I'm also relatively new to web dev, so take this with a grain of salt and set me straight if I am wrong!

**Comment by a-h** [comment]
Hi folks.

Quick update on this. The conversation carried on in #838, and has resulted in a PR: https://github.com/a-h/templ/pull/1038

With this change, there's no need to mess about with data attributes, call a function directly and pass the server-side value to the function as a JavaScript object.

```templ
<button onClick={ templ.JSFuncCall("alert", "Hello, World!") }>Click me</button>
```

This results in:

```html
<button onClick="alert('Hello, World!')">Click me</button>
```

To define a custom handler, and pass data to it:

```templ
@onceHandle.Once() {
	<script type="text/javascript">
		function customAlert(msg, date) {
			alert(msg + " " + date);
		}
	</script>
}
<button onClick={ templ.JSFuncCall("customAlert", "Hello, custom alert 1: ", time.Date(2020, 1, 1, 0, 0, 0, 0, time.UTC)) }>Click me</button>
```

To render a call to a JS function, passing your custom data drop this anywhere.

```templ
@templ.JSFuncCall("customAlert", "Runs on page load", time.Date(2020, 1, 1, 0, 0, 0, 0, time.UTC))
```

This renders a script tag with the function call in it:

```html
<script>
customAlert("Runs on page load", "2020-01-01T00:00:00Z")
</script>
```

If you want to (dangerously, because there's no XSS protection) pass arbitrary values, you can pass an expression:

```templ
<button onclick={ templ.JSFuncCall("onClickEventHandler", templ.JSExpression("event"), "1234") }>Pass event handler</button>
```

That will result in:

```html
<button onclick="onClickEventHandler(event, '1234')">Pass event handler</button>
```

That's in addition to the `templ.JSON` and `templ.JSONScript` functions that can be used to put JSON values into element attributes where tools like alpine.js can access them (e.g. the `x-data` attribute in alpine).

**Comment by dimmerz92** [comment]
For those of us not able to use inline attributes for CSP reasons, how can we get data into a script tag on the server side?

**Comment by a-h** [comment]
https://templ.guide/syntax-and-usage/script-templates/#pass-server-side-data-to-the-client-in-a-script-element

Check out `.WithNonce`.

**Comment by dimmerz92** [comment]
Yeah, already using with nonce for other script tags that don't require any go variables.

Was just hoping there'd be a way to get go variables directly into script tags without having to do the workaround.

**Comment by a-h** [comment]
I think templ could do something similar to what stdlib templates do, however, it does of course mean that there's a chance of producing invalid JS, as per this example: https://repeatit.io/#/share/eyJ0ZW1wbGF0ZSI6IjxkaXY+XG48c2NyaXB0IHR5cGU9XCJ0ZXh0L2phdmFzY3JpcHRcIj5cbnZhciB4ICdUaGUgbWVzc2FnZSBpcyBcInt7IC5tc2cgfX1cIidcbnZhciB5IHt7IC5tc2cgfX1cbnZhciB6ICd7eyAubXNnIH19J1xue3sgLm1zZzIgfX1cbjwvc2NyaXB0PlxuPC9kaXY+IiwiaW5wdXQiOiJtc2c6IFwiSGVsbG8gJyBXb3JsZCFcIlxubXNnMjogPC9zY3JpcHQ+IiwiY29uZmlnIjp7InRlbXBsYXRlIjoiaHRtbCIsImZ1bGxTY3JlZW5IVE1MIjpmYWxzZSwiZnVuY3Rpb25zIjoiaGVsbSIsIm9wdGlvbnMiOlsibGl2ZSJdLCJpbnB1dFR5cGUiOiJ5YW1sIn19

But... I think doing that is another issue - I think this one has ran its course.
